llms/mlx_lm/sample_utils.py

```diff
@@ -196,8 +208,15 @@ def top_p_sampling(logits: mx.array, top_p: float, temperature: float) -> mx.arr
         0,
     )
 
-    sorted_tokens = mx.random.categorical(mx.log(top_probs), axis=-1)[:, None]
-    return mx.take_along_axis(sorted_indices, sorted_tokens, axis=-1).squeeze(1)
+    # Create a mapping to rearrange back to original indices
+    # Use argsort of sorted_indices to get the inverse permutation
+    inverse_indices = mx.argsort(sorted_indices, axis=-1)
+
+    # Rearrange top_probs back to original order
+    original_order_probs = mx.take_along_axis(top_probs, inverse_indices, axis=-1)
+
+    # Convert back to logits and return
+    return mx.log(mx.where(original_order_probs > 0, original_order_probs, 0))
```

Comment:
We shouldn't set the `0` probs to `0` post log. They should be `-inf` in log space. Did you try it without the `mx.where`? I think it should work if you just remove that.. but if not.. there might be a downstream issue.